samples/weather_function/src/main/kotlin/org/example/weather_func/curl.kt

```diff
@@ -0,0 +1,89 @@
+package org.example.weather_func
+
+import org.example.weather_func.Event
+import platform.posix.size_t
+
+import curl.curl_easy_setopt
+import curl.CURLOPT_URL
+import curl.CURLOPT_HEADERFUNCTION
+import curl.CURLOPT_HEADERDATA
+import curl.CURLOPT_WRITEFUNCTION
+import curl.CURLOPT_WRITEDATA
+import curl.curl_easy_cleanup
+import curl.curl_easy_init
+import curl.CURLE_OK
+import curl.curl_easy_strerror
+import curl.curl_easy_perform
+
+import kotlinx.cinterop.staticCFunction
+import kotlinx.cinterop.COpaquePointer
+import kotlinx.cinterop.CPointer
+import kotlinx.cinterop.StableRef
+import kotlinx.cinterop.asStableRef
+import kotlinx.cinterop.ByteVar
+import kotlinx.cinterop.readBytes
+
+/**
+ * Provides basic HTTP client functionality through the libCurl library.
+*/
+internal class CUrl(val url: String) {
+    private val stableRef = StableRef.create(this)
+    private val curlObj = curl_easy_init()
+    val header = Event<String>()
+    val body = Event<String>()
+
+    init {
+        val header = staticCFunction(::headerCallback)
+        val writeData = staticCFunction(::writeCallback)
+
+        // Setup Curl.
+        curl_easy_setopt(curlObj, CURLOPT_URL, url)
+        curl_easy_setopt(curlObj, CURLOPT_HEADERFUNCTION, header)
+        curl_easy_setopt(curlObj, CURLOPT_HEADERDATA, stableRef.asCPointer())
+        curl_easy_setopt(curlObj, CURLOPT_WRITEFUNCTION, writeData)
+        curl_easy_setopt(curlObj, CURLOPT_WRITEDATA, stableRef.asCPointer())
+    }
+
+    fun fetch() {
+        // Have Curl do a HTTP/HTTPS request and store the response (status).
+        val response = curl_easy_perform(curlObj)
```

Comment:
I'm curious (and new to Kotlin).. is there no better way to do this than linking to curl? I've never really seen that done in a higher-level language - there is normally some kind of http package in the stdlib or ecosystem.

Reply:
No, at least before Kotlin Native 0.8. However with the recent Kotlin Native [release](https://blog.jetbrains.com/kotlin/2018/07/kotlinnative-v0-8-released/) there is a opportunity to use the Kotlin Native HTTP client library (a port of the Ktor ver from the JVM side) instead of Curl. Kotlin JVM doesn't have this issue since there are some well established HTTP client libraries available ([OkHttp](http://square.github.io/okhttp/), [Retrofit](http://square.github.io/retrofit/), [Apache HttpClient](https://hc.apache.org/httpcomponents-client-ga/) etc).

If there are some people involved in the development of the new Kotlin Native HTTP Client library, who are reading this then please get in contact. Interested in migrating from Curl to the new library.

Reply:
@alexellis - Received a response about the Ktor Linux port [issue](https://github.com/ktorio/ktor/issues/479). The port is being taken under consideration for the near future. In the meantime Curl remains the HTTP client library to use for this sample even though it isn't a ideal fit  :(.